[BEAM-9008] Add CassandraIO readAll method

vmarquez opened the pull request:
Edit: This PR adds a readAll method that returns a PTransform<PCollection<Read<A>>, PCollection<A>> to allow for more fine grained querying of a Cassandra compatible databases (Cassandra, Scylla etc).  Since each Read<A> allows one to specify a Query and/or RingRange, a pipeline does not have to do a full table scan to retrieve data, it can now read a subset using this pattern.  

Because Read<A> allows the user to specify a Set<RingRange>, this allows even more flexibility, in that a user can decide to put N number of RingRanges (which each could be representing individual keys) in a single Read<A>, this allows one to *linearize* queries, which is very important when doing a large number of individual queries so as to not overload single shards or cores.   


_Old comment for posterity, not relevant to the current redesign after working with @iemejia: This PR adds a readAll method that returns a `PTransform<PColl<List<RingRange>>, PCollection<T>>` to allow for parallel querying of a small subset of a Cassandra databse.  In addition, it refactors the current method `read` that returns a `PTransform<PBegin, PColl<T>>` to be a ParDo based one, so we can share code between read and readAll.  
Note: One thing I wasn't sure about was how to abstract over AutoValue classes.  I'm much more familiar with Scala than Java and haven't used AutoValue before, so I wasn't sure if there was a better way than what I did with having a `getCassandraConfig` on both the static Read class as well as the ReadAll class.  Don't think I can use an interface since I don't want the methods public.  Open to suggestions here._ 

Thanks in advance for being patient with a large change and please let me know if I can change/add anything!  

R:@iemejia 
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

suztomo commented:
Would you ensure the change passes ` ./gradlew -p sdks/java check`? I see it fails with format violations:

```
suztomo-macbookpro44:vmarquez-beam suztomo$ git log -1
commit d85e6b95d64664af37148c792b35ff02abfd90f5 (HEAD -> feature/BEAM-9008/cassandraio_readall, origin/feature/BEAM-9008/cassandraio_readall)
Author: Vincent Marquez <vincent.marquez@ytel.com>
Date:   Wed Dec 18 16:48:52 2019 -0800

    implementing CassandraIO.readAll
suztomo-macbookpro44:vmarquez-beam suztomo$ ./gradlew -p sdks/java check
Configuration on demand is an incubating feature.

...

* What went wrong:
Execution failed for task ':sdks:java:io:cassandra:spotlessJava'.
> The following files had format violations:
      sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java
          @@ -184,31 +184,33 @@
           
           ····LOGGER.info("Insert·records");
           ····String[][]·scientists·=·{
          -······new·String·[]·{"phys",·"Einstein"},
          -······new·String·[]·{"bio",·"Darwin"},
          -······new·String·[]·{"phys",·"Copernicus"},
          -······new·String·[]·{"bio",·"Pasteur"},
          -······new·String·[]·{"bio",·"Curie"},
          -······new·String·[]·{"phys",·"Faraday"},
          -······new·String·[]·{"math",·"Newton"},
          -······new·String·[]·{"phys",·"Bohr"},
          -······new·String·[]·{"phys",·"Galilei"},
          -······new·String·[]·{·"math",·"Maxwell"},
          +······new·String[]·{"phys",·"Einstein"},
          +······new·String[]·{"bio",·"Darwin"},
          +······new·String[]·{"phys",·"Copernicus"},
          +······new·String[]·{"bio",·"Pasteur"},
          +······new·String[]·{"bio",·"Curie"},
          +······new·String[]·{"phys",·"Faraday"},
          +······new·String[]·{"math",·"Newton"},
          +······new·String[]·{"phys",·"Bohr"},
          +······new·String[]·{"phys",·"Galilei"},
          +······new·String[]·{"math",·"Maxwell"},
           ····};
           ····for·(int·i·=·0;·i·<·NUM_ROWS;·i++)·{
           ······int·index·=·i·%·scientists.length;
          -······String·insertStr·=·String.format(
          -··········"INSERT·INTO·%s.%s(person_department,·person_id,·person_name)·values("
          -··············+·"'"·+·scientists[index][0]·+·"',·"
          -··············+·i
          -··············+·",·'"
          -··············+·scientists[index][1]
          -··············+·"');",
          -··········CASSANDRA_KEYSPACE,
          -··········CASSANDRA_TABLE);
          -········LOGGER.error("Error·with·str·=·"·+·insertStr);
          -······session.execute(insertStr
          -··········);
          +······String·insertStr·=
          +··········String.format(
          +··············"INSERT·INTO·%s.%s(person_department,·person_id,·person_name)·values("
          +··················+·"'"
          +··················+·scientists[index][0]
          +··················+·"',·"
          +··················+·i
          +··················+·",·'"
          +··················+·scientists[index][1]
          +··················+·"');",
      ... (102 more lines that didn't fit)
  Violations also present in:
      sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/QueryFn.java
      sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraConfig.java
      sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java
  Run 'gradlew spotlessApply' to fix these violations.
```

iemejia commented:
Arrghh tests for this one did not start, for that reason you had not seen the failures that @suztomo mentioned.
You can auto fix the code format issues by running `./gradlew :sdks:java:io:cassandra:check spotlessApply`.
I will start letting some review comments in the meantime

vmarquez commented:
@suztomo @iemejia Thank you for pointing that out, I was not aware of the formatting issues.  Maybe after this MR I can add some of that to the contributing document. :)  I fixed the style and formatting issues, and will await other change requests.

iemejia commented:
Arrghh unit tests did not run because of the issues with jenkins this week, running again.

iemejia commented:
Working still in the review sorry for the delay @vmarquez

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -325,7 +359,58 @@ private CassandraIO() {}
+        if (isMurmur3Partitioner(cluster)) {
+          LOG.info("Murmur3Partitioner detected, splitting");
+
+          List<BigInteger> tokens =
```
For maintenance coherence in DoFn based IOs we follow a pattern where the first part Creates the required object to query (the one I mentioned above), the second splits the query into multiples queries for each partition via a `SplitFn extends DoFn` and then with the result of such splits we do a read equivalent to current `QueryFn`. Can you refactor this segment to behave like that (with the query object I mentioned above).

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -608,121 +470,32 @@ static double getRingFraction(List<TokenRange> tokenRanges) {
+   * range.
+   */
+  @VisibleForTesting
+  static class TokenRange {
```
We don't need this one anymore I think we can remove this class now as well as: `CassandraIO.getRingFraction` and `CassandraIOTest.testRingFraction`. Can you remove these too please.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/RingRange.java:
```diff
@@ -17,10 +17,11 @@
 
 /** Models a Cassandra token range. */
-final class RingRange {
+final class RingRange implements Serializable {
```
Actually this should probably be done not for this class but for the `CassandraQuery` object

iemejia commented on sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java:
```diff
@@ -312,6 +298,52 @@ public void testRead() throws Exception {
 
+  @Test
+  public void testReadAll() {
+
```
nit: remove space

iemejia commented on sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java:
```diff
@@ -387,6 +420,7 @@ public Mapper apply(Session input) {
 
     @Override
     public Iterator map(ResultSet resultSet) {
+
```
nit: remove space

iemejia commented on sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java:
```diff
@@ -466,40 +500,9 @@ public void testCustomMapperImplDelete() {
 
-  @Test
-  public void testSplit() throws Exception {
-    PipelineOptions options = PipelineOptionsFactory.create();
```
Once we have the SplitFn we should probably test this, actually we had a lot of historic issues with Splitting on this connector so please pay attention to ensure we don't introduce regressions.

vmarquez replied:
Hmm, if you know of relevant issues that we previously had, lmk, I will also look through the issue tracker to see if there were previous bugs.

iemejia commented:
Btw, don't hesitate to ask questions or discuss here if you have any doubts before starting.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/QueryFn.java:
```diff
@@ -0,0 +1,89 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class QueryFn<T> extends DoFn<Iterable<RingRange>, T> {
```
I don't think we want this flattened here, the way I'm using readAll currently is I am hashing into groups of queries based on actual shard, which is owned by a core (in ScyllaDB), so grouping queries gives finer grained control over how much load we put on a core.  Someone using cassandra could use a similar split idea for making sure to limit the number of concurrent TokenRanges owned by the same node in Cassandra.

vmarquez commented:
Thanks for taking a look, I've pushed a new commit (I won't be squashing anymore until we finalize everything) that hopefully addresses all the minor style issues.  LMK if I missed anything. 

I do want to make sure we keep the cassandraIO 'idiomatic' to the rest of the IO connectors, but I don't think modeling this after the SOLR one will work.  For one thing, if we want to share the `ReadFn` class between both Read and ReadAll, it means we have to have some way of having both use it and pass in 'connection' information, which we can't do if the signature of ReadFn is `ReadFn extends DoFn<ReadAll<A>, A>`.  I think another class to look at for something that has both Read and ReadAll PTransforms is the SpannerIO, which is modeled similarly to how I did it here (though not exactly). 

https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java#L315

They have a configuration class there is public (which we can't do since we want to keep backwards compatibility with the current way `Read` works), it has two different PTransforms, `Read` uses `ReadAll` internally, etc.  

I do think instead of taking a collection of RingRanges, taking some sort of 'Query' object makes sense, and the idea that it doesn't have to tie in to the actual connection means we can split up the CassandraConfig class.  Thoughts on that?

iemejia commented:
Sorry again for taking a bit to respond, my review queue ended up being longer than usual this week.

I agree on your intuition on letting the component idiomatic, I just took a look at SpannerIO and it is indeed aligned with my first idea `ReadOperation` == `Query`, (I call it query following the naming of `HbaseQuery`). Naming is less important than the fact that this object ‘should have’ only parameters related to the Read/Query and not to the connections. So I agreed 100% on that.

However my intuition changed a bit when I started to think that `ReadOperation/Query` in every case is a subset of Read. Read = Connection + ReadOperation/Query, so why we need ReadOperation at all? We can have just the normal Read but if you set up explicitly the RingRage as an extra parameter well we don’t split and just read from it. SolrIO creates an artificial KV to put that ‘query’ in the side of the normal Read, but for example DynamoDBIO does the read based on an attribute and you could have an ultra simple ReadAll.

```java
  public static class ReadAll<T> extends PTransform<PCollection<Read<T>>, PCollection<T>> {
    @Override
    public PCollection<T> expand(PCollection<Read<T>> input) {
      return input.apply(ParDo.of(new ReadFn<>()));
    }
  }
```

In the end `ReadAll` is more a kind of `Read Split/Partition` (without trying to split), so the only extra thing we need to have for that is the info that allows us to read the specific partition: `ReplicaInfo` in `Solr`,  `segmentId` in `DynamoDB` and `RingRange` in `Cassandra`. Also having the full Read would allow users to read from different systems once the user creates the PCollection<Read> so this is a huge reusability advantage.

Do you see any other tradeoffs, for me it looks quite straight forward and simpler, we won't need to maintain additional objects. Maybe worth trying it.

vmarquez commented:
@iemejia I think that could work, thanks for your patience as I try to understand what you're thinking.  Some questions:

1.  If our `ReadFn<A> extends DoFn<Read<A>, A>` and the only way we have connection information is from the actual read, that means we need to re-establish a DB connection for each  batch of queries we run?  

2. How would that work for the end user of a `PTransform<PCollection<Read<A>, PCollectoin<A>>`?  Here is what I did in the test and would wand to document how end users could generate 'queries',
https://github.com/apache/beam/pull/10546/files#diff-8ba4ea3b09d563a67e29ff8584269d35R499

Would we instead want to return a `PCollection<Read<A>>` by using something like `return CassandraIO.read().withRingRange(new RingRange(start, finish))`? 

3.  Originally I had wanted to have the ReadFn operate on a *collection* of 'query' objects to ensure a way to enforce linearizability with our queries (mainly so we don't oversaturate a single node/shard).  Currently the groupBy function a user passes in operates on the `RingRange` object, would we keep it that way and just, under the hood, allow for a single `Read<A>` to hold a collection of RingRanges?

iemejia commented:
> @iemejia I think that could work, thanks for your patience as I try to understand what you're thinking. Some questions:

No, thanks to you who has been the patient one during this discussion.

>     1. If our `ReadFn<A> extends DoFn<Read<A>, A>` and the only way we have connection information is from the `Read<A>` passed in to the processElement, that means we need to re-establish a DB connection for each  batch of queries we run?  As in, the connection would be established in the `processElement` method and could not be in `setup` method?

Yes exactly this will make the method simpler and the cost of starting a connection gets amortized by the processElement producing multiple outputs from a single connection.

>     2. How would that work for the end user of a `PTransform<PCollection<Read<A>, PCollection<A>>`?  Here is what I did in the test and would wand to document how end users could generate 'queries',
>        https://github.com/apache/beam/pull/10546/files#diff-8ba4ea3b09d563a67e29ff8584269d35R499
>        Would we instead want to return a `PCollection<Read<A>>` by using something like `return CassandraIO.read().withRingRange(new RingRange(start, finish))`?  If we do that however, we'd need to do the `withHosts` and all the other connection information, no?  The other option is establishing one `ReadAll` PTransform that maps over the `Read<A>` input and enriches the db connection information?

You have a point here!. We need `class ReadAll extends PTransform<PCollection<Read<A>, PCollection<A>>` and there we read as intended with `ReadFn`. You would have to modify however the `expand` of `Read` to do `input.apply(Create.of(this)).apply(CassandraIO.readAll())` where `ReadAll` should expand into `input.apply(ParDo.of(splitFn)).apply(Reshuffle).apply(Read)` users should deal with building the PCollection of `Reads` before passing that collection to `ReadAll`.

>     3. Originally I had wanted to have the ReadFn operate on a _collection_ of 'query' objects to ensure a way to enforce linearizability with our queries (mainly so we don't oversaturate a single node/shard).  Currently the groupBy function a user passes in operates on the `RingRange` object, would we keep it that way and just, under the hood, allow for a single `Read<A>` to hold a collection of RingRanges?

If I understand this correctly this is covered by following the Create -> Split -> Reshuffle -> Read pattern mentioned above (in the mentioned IOs). So Split is the one who will generate a collection of `Read`s for each given `RingRange` then we use Reshuffle to guarantee that reads are redistributed and finally each read request is read by one worker.

Hope this helps, don't hesitate to ask me more questions if still. I will try to answer quickly this time.
Hope this helps

iemejia commented:
Forgot to mention that in the above comment that in the Split function you have to split in every case save if the user provided a specific RingRange to read from.

vmarquez commented:
@iemejia I have started working on a different implementation of ReadAll that extends `PTransform<PCollection<Read<T>>, PCollection<T>>`, however, I have thought about it more and I am thinking that it is bad to not be able to establish a connection in `setup` of the `ReadFn` class.  Just this week at work I had to do some work involving generating hundreds of thousands of queries (`RingRanges`), and the vast majority of them returned no or almost no data!  The data was sparse and never the less I still needed to perform the queries, so having to establish a new connection per `RingRange` in the `processElement` method would have slown things down a lot.  

I understand the desire to have simpler code, but I think the performance impact would be noticable if we only keep connection information in the `Read<T>` that comes into the `processElement` method.   Do you still want to proceed with that?

iemejia commented:
Cassandra is extremely slow opening new connections however the average case would be that for each `Read` there would be many `T` outputs which in principle would amortize the slow connection time. Couldn't the user in the specific case you mention of many Reads generate queries for RingRanges that match more 'regions' and achieve the same result?

If this is really ends up being a problem one eventual workaround would be to batch requests and create some sort of pool of connections but hopefully this won't be needed.

vmarquez commented:
> Couldn't the user in the specific case you mention of many Reads generate queries for RingRanges that match more 'regions' and achieve the same result?

 I don't think so, a user might be querying specific partition keys (only one key that returns multiple results) that are not contiguous some of which have data and many of which may have no data.

iemejia commented:
I think this is needing a rebase. Please ping me when this is ready for review. I am eager to do so hoping this can make it into Beam 2.21.0

iemejia commented:
I was also thinking about the potential connection issue and I think we can try to tackle it by creating some sort of static connection pool and we id the connections by their server/config and reuse it if possible. We can work on this in this PR or on a follow up, as you prefer.

vmarquez commented:
Hi @iemejia I've been busy with work and life but was finally able to this just about finished up!  To get around the connection issue (not caching the connection was causing the Cassandra tests to run for ~ten minutes!), by passing in a ReadAll<T> to the ReadFn, we're able to initiate a connection in the setup method, but we can still 'dynamically' use the passed in Read<T> to generate specific queries or query ranges.  

As for a more advanced connection pooling, I'd prefer to get this merged in and then perhaps work on an additional PR.  

There are some minor conflicts but I wasn't sure how you wanted to handle that, do you prefer I rebase against master (? that might then cause issues with Github's UI seeing other folks commits in my changelist), or should I merge master into this then another commit(which makes later squashing a bit harder).  Advice on that would be appreciated!  Thanks again for your guidance and help on this. I like the new design and the flexibility to modify the queries as well as ring ranges is pretty neat.

vmarquez commented:
@iemejia I've fixed the merge conflicts now with a rebase and I'm ready for further review. Thanks in advance!

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
retest this please

iemejia commented:
bang

iemejia commented:
retest this please

iemejia commented:
retest this please

vmarquez commented:
@iemejia is that a spurious failure or did something I do break the Flink test?  I tested locally and all seems to work... LMK if you need anything from me.

iemejia commented:
Yes looks like a flaky test there is already a ticket tracking it https://issues.apache.org/jira/browse/BEAM-9534

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -318,15 +354,84 @@ private CassandraIO() {}
+          SplitGenerator splitGenerator =
+              new SplitGenerator(cluster.getMetadata().getPartitioner());
+
+          List<Read<T>> splits =
```
We should produce a PCollection<Read<T>> here instead and this splitting logic should be moved into the SplitFn function.

vmarquez replied:
This is more or less leaving the splitting logic the same as it was before, is there a reason we need to change it now?

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -318,15 +354,84 @@ private CassandraIO() {}
+      }
+    }
+
+    private static class SplitFn<T> extends DoFn<List<RingRange>, Read<T>> {
```
`private static class SplitFn<T> extends DoFn<Read<T>, Read<T>>`
Notice that in the second Read you should setup the `RingRange` so each individual Read can then be Read by a DoFn that knows how to read `Read` specifications with a RingRange.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -318,15 +354,84 @@ private CassandraIO() {}
+    private static class SplitFn<T> extends DoFn<List<RingRange>, Read<T>> {
+      @ProcessElement
+      public void processElement(@Element List<RingRange> input, OutputReceiver<Read<T>> output) {
+        for (RingRange rr : input) {
```
The split logic that is above should be here.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1170,4 +887,344 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+  public abstract static class ReadAll<T> extends PTransform<PCollection<Read<T>>, PCollection<T>> {
+
+    @Nullable
+    abstract ValueProvider<List<String>> hosts();
```
This transform should not have any attribute / with method, those come from the Read in the input PCollection.

vmarquez replied:
For ReadAll to be useable we need to establish connection information in the `setup` phase of the QueryFn, so we'll need all this before a `Read<T>` comes in.   I'd also like to keep the API between ReadAll and Read similar so it's easy for the end user to switch a `Read<T>` with a `ReadAll<T>`, and just operating on a different type.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1170,4 +887,344 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+      checkArgument(entity() != null, "withEntity() is required");
+      checkArgument(coder() != null, "withCoder() is required");
+      checkArgument(groupingFn() != null, "GroupingFn OR splitCount must be set");
+      try (Cluster cluster =
```
The key point here is to apply for every read the split function and produce subsequent reads with RingRanges that then are Reshuffled and passed to a ParDo with the ReadFn function.

vmarquez replied:
Not sure I understand what you mean here.  I'm doing a shuffle by key here, `FlattenGrouped` here is what I called it, following the other connectors general pattern.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/ReadFn.java:
```diff
@@ -0,0 +1,109 @@
+  private final ReadAll<T> readAll;
+
+  public ReadFn(ReadAll<T> readAll) {
+    this.readAll = readAll;
```
You don't need this ReadAll ref here, Read is enough, ReadAll is the one that delegates to a normal Read.

vmarquez replied:
You need this here so we can set up a connection with Setup, this won't work without it.

vmarquez replied:
If you want I can spend more time getting performance data, but without caching the connection on setup I saw more than two orders of magnitude slowdown on just the `CassandraIOTest`

iemejia replied:
Caching connections is something we can do in a subsequent PR. Let's get first right the API and then optimize that part. We can cache connections at the class level on the `ReadFn` by creating some size limited pool of connections that we identify by some combination of parameters and we reuse if possible.

vmarquez replied:
@iemejia the tests don't even finish running *locally* after ten minutes when running `CassandraIOTest` if we don't cache the connection.  I have already tried doing it without establishing a connection in `setup`, it's not workable.  The CassandraIOTest returns over *200* partitions to return *10* rows.  That means it has to establish over 20 connections and tear them down per row...

iemejia commented:
Just for ref I opened the PR for SolrIO #11357 that introduces ReadAll there.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraConfig.java:
```diff
@@ -0,0 +1,88 @@
+  abstract ValueProvider<List<String>> hosts();
+
+  @Nullable
+  abstract ValueProvider<String> query();
```
This class isn't used anymore, I will delete this.  All of the relevant information for a connection is now on the Read being passed in to ReadAll.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -391,36 +393,57 @@ static Cluster getCluster(
+    }
 
-          return input.apply("Creating splits", Create.of(splits)).apply("readAll", readAll);
+    private static class SplitFn<T> extends DoFn<Read<T>, Read<T>> {
```
This looks perfect!

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -391,36 +393,57 @@ static Cluster getCluster(
+            .apply(ParDo.of(new SplitFn<T>()))
+            .setCoder(SerializableCoder.of(new TypeDescriptor<Read<T>>() {}))
+            .apply(Reshuffle.viaRandomKey())
+            .apply(readAll);
```
Now we need to tackle this in two parts maybe, one is to implement the read with a `ReadFn` like method  and as a next step to get rid of all the methods on `ReadAll` to simplify it to its core.

iemejia commented:
retest this please

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1170,4 +898,44 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+      return input
+          .apply("shuffle", Reshuffle.viaRandomKey())
+          .apply("read", ParDo.of(new ReadFn<>()))
+          .setCoder(this.coder());
```
coder should come from the input read

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1170,4 +898,44 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+    abstract ReadAll.Builder<T> builder();
+
+    /** Specify the {@link Coder} used to serialize the entity in the {@link PCollection}. */
+    public ReadAll<T> withCoder(Coder<T> coder) {
```
I am not sure if we need this, I assume the coder comes form the `Read<T> input.

vmarquez replied:
I think we do, but am not sure.  You have to call `setCoder` on the PCollection itself, so we don't have access to the `Read<T>` at a point when we also have access to a single `Read<T>` (they are only supplied in the `ReadFn` which can't call `setCoder` on the returned PTransform).  Is my thinking correct?  I could be unaware of a different way to do this.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -326,7 +371,78 @@ private CassandraIO() {}
+                read.password(),
+                read.localDc(),
+                read.consistencyLevel())) {
+          if (isMurmur3Partitioner(cluster)) {
```
Oops I did forget that, I will add. Good catch.

vmarquez commented on sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java:
```diff
@@ -480,66 +527,22 @@ public void testCustomMapperImplDelete() {
     return result.all();
   }
 
+  // TEMP TEST
```
Ooops I think this was left in from when I was experimenting with something, I will remove.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -326,7 +371,78 @@ private CassandraIO() {}
+                .forEach(
+                    rr ->
+                        outputReceiver.output(
+                            CassandraIO.<T>read()
```
I think we still need to create a new `Read<T>` because the `SplitGenerator` returns a `List<List<RingRange>>`, so each of the outer list will be a different Read, the inner List will be set to the RingRange.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -370,384 +488,16 @@ private CassandraIO() {}
-      for (List<RingRange> split : splits) {
-        List<String> queries = new ArrayList<>();
-        for (RingRange range : split) {
-          if (range.isWrapping()) {
```
Hm, let me think about this.

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -326,7 +371,78 @@ private CassandraIO() {}
+
+      return input
+          .apply(Create.of(this))
+          .apply(ParDo.of(new SplitFn()))
```
> Move the Split as the first step of the ReadAll expansion so non advanced users (those who do not specify RingRange manually could get their code 'partitioned' correctly.

I am not sure I understand what you mean here, could you explain more?

vmarquez commented:
@iemejia Apologies for the long delay, I finally had time to deal with the big merge and I believe I have addressed all the issues. Maybe a few nits I missed.  I also realized I needed to make it so readAll could work on a PCollection of reads even without a ringrange, in case I wanted to just have various queries, so now there's even more flexibility in how we can use readAll.  I added an additional test to show that as well.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1281,4 +1032,44 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
```
Move this up before the `withCoder` method

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -1281,4 +1032,44 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+    @Nullable
+    abstract Coder<T> coder();
+
+    abstract ReadAll.Builder<T> builder();
```
Remove `ReadAll.`

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -373,7 +419,86 @@ private CassandraIO() {}
+
+      return input
+          .apply(Create.of(this))
+          .apply("Split", ParDo.of(new SplitFn()))
```
Can you please move this Split into the ReadAll expand method. When not specified RingRanges like in the `read()` case we should split them on the `ReadAll` expansion. Of course if RingRanges are specified we will probably ignore recalculating them.

vmarquez replied:
I don't think it makes sense to move to the readAll...  The readAll takes a Read<>, so on encountering each Read<>, we want to check if there isn't a ringRange or Query set, and then expand it?  I think it is much cleaner here, and I think this is where you originally suggested the split happen?

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/ReadFn.java:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  private static <T> boolean reuseCluster(Read<T> readA, Read<T> readB) {
+    return readA != null
```
This is still a bit messy, don't you think we can get something similar by using something like connection pooling and delegate this complexity to the driver (where it should reside)?
https://docs.datastax.com/en/developer/java-driver/2.1/manual/pooling/

Notice that I am not familiar with this but after a quicklook seems to be worth the look.

vmarquez replied:
I didn't address this yet, wanted your thoughts on it.  I agree it can be improved. 

The driver will handle Connection *pooling*, (handling N number of connections per cassandra host and sharing them in a thread safe way), but we still need to initiate a `Session` per Cassandra instance, and maintain a reference to the `Session` to keep those connections in scope (under the hood).  

So we could use something as simple as a `Map<?, Cluster>` and `Map<?, Session>`, but remember we have to do both, since each `Read<>` might be hitting an entirely different cassandra cluster.   If we wanted something fancier, we could try a guava LRU or something on the off chance the user is initiating hundreds(thousands/millions?) of different Session objects.

iemejia commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/ReadFn.java:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  /*
+  private static String generateRangeQuery(Read<?> spec, String partitionKey) {
```
Remove if unused

vmarquez commented on sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java:
```diff
@@ -143,6 +148,36 @@ private CassandraIO() {}
   }
 
+  /** Get a Cassandra cluster using hosts and port. */
+  static Cluster getCluster(
```
@iemejia this was done accidentally after the merge.  I will remove it.

iemejia commented:
It is great to know we can still test this in a good scale in practice so a good motivation to finish it. I expect to be a bit more available next week so let's try to push it to the end.

vmarquez commented:
@iemejia one more thing, regarding the split testing, on the original PR I had removed it, but once we decided on simplifying things to the architecture of `Read<T>` doing the splits and then passing to `ReadAll<T>`, I kept the split logic exactly the same and put the test back.

TheNeuralBit commented:
Is this ready for @iemejia to review again?

vmarquez commented:
@TheNeuralBit Yes, thanks for checking.  @iemejia and I were working in his fork with some changes going back and forth, i've merged them here and this should be the up to date PR ready for review again. It is large, so happy to have anyone else take a look as well if you have time.

aromanenko-dev commented:
Kind ping on this since it hasn't been updated since February. What is a state of this PR?

vmarquez commented:
@aromanenko-dev I am still waiting any final change requests, but would like to get it merged if there are no further changes.

vmarquez commented:
@iemejia @aromanenko-dev @TheNeuralBit 

Hello, just checking in to see if this could get reviewed and/or merged anytime soon, it has been open quite a while.  Please let me know if there's anything I can do to help, I don't mind writing more documentation or answering any questions.

vmarquez commented:
Any reason why that particular test would fail? Passing locally for me.

aromanenko-dev commented:
Run Java PreCommit

aromanenko-dev commented:
@vmarquez Sorry for delay with answer, I don't think that this fail is related to your changes. 
I'd leave @iemejia to finish this review since he already was on this.

aromanenko-dev commented:
Run Java PreCommit

vmarquez commented:
@aromanenko-dev @iemejia I haven't heard anything in months,  still interested in this contribution?

aromanenko-dev commented:
@vmarquez Sorry for delay with an answer, this contribution is still very valuable! I hope @iemejia can get back to this PR soon. Otherwise, I'll take it over. 

ping @iemejia

vmarquez commented:
@aromanenko-dev It's been a couple weeks, is there anything I can do to make this easier to review or help this process out?  LMK if you want more docs/comments/a walk through or anything. 

I can be available on the slack as well to chat in realtime if that would help.

aromanenko-dev commented:
@vmarquez I'm sorry for delay with a response - I was on holidays and quite busy with Beam Summit in the last weeks. I need some time to refresh my knowledge of this PR and I'll get back to you after.

aaltay commented:
> @vmarquez I'm sorry for delay with a response - I was on holidays and quite busy with Beam Summit in the last weeks. I need some time to refresh my knowledge of this PR and I'll get back to you after.

@aromanenko-dev - kind ping to see if you have some time.

aromanenko-dev reviewed (commented):
@vmarquez Thanks for your work and sorry for delay again. 

I finally managed to take a look on the code and the related discussion to get the main idea of this PR and how it was evolved along the time (it has beed changing pretty well since the beginning). So I expect that @vmarquez and @iemejia came to the final version  of design which was implemented and it should be a good point. Please, let me know if I'm mistaken about this.

Though, I have several questions (sorry if I missed something and it was already discussed before).

1. Can we expect any breaking user API changes?
2. Does this PR contain all necessary tests to test a new functionality and corner cases?
3. @vmarquez Did you try to run it on a real distributed cluster with a real data? If yes, could you provide any details on this (if possible)?
4. Do we expect any performance degradation (or improvement!) for `CassandraIO.read()` since a Read part was mainly rewritten? It would be great to compare.
5. Have been all PR's comments addressed? If not, what is missing?

vmarquez commented:
> Can we expect any breaking user API changes?

No breaking API changes.

>Does this PR contain all necessary tests to test a new functionality and corner cases?

Yes, the two tests I added test both using RingRanges and Queries.  

>@vmarquez Did you try to run it on a real distributed cluster with a real data? If yes, could you provide any details on this (if possible)?


Sure!  The company I'm working for now  has been running this current version (of my  fork) in production for six months.   We've used it in various dataflow pipelines on a cluster of 5 Scylla servers.   For batch runs we've gotten up to ~3M  elements per second running with hundreds of dataflow worker nodes, for streaming we've only used it on a much smaller sacle, but been doing up to hundreds per second on two nodes. 

>Do we expect any performance degradation (or improvement!) for CassandraIO.read() since a Read part was mainly rewritten? It would be great to compare.

For improvements, I think this basically makes some things possible that just weren't before.  If you have billions of rows it's almost untenable to have to read the entire DB into ram and then filter if you only want a subset.  With readAll you can programatically generate thousands or tens of thousands of queries and send it to readAll instead.  

I don't expect any performance degredation (after I make one final change Ismeal has asked for): Previously the group of ring ranges that were queried were done with async calls.   This doesn't make a lot of sense since in essence it 'undoes' the splitting of the ring ranges (why not just make more splits and have more sync calls?)  

One of the ways we are able to achieve such good performance for large batch jobs with this current change is due to being able to linearize the queries and not overload a specific shard (or core, in Scylla's case).  

>Have been all PR's comments addressed? If not, what is missing?

https://github.com/vmarquez/beam/blob/feature/BEAM-9008/cassandraio_readall/sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java#L418

@iemejia wanted me to flatten the results here so to mitgiate the change from async to sync DB calls:  The idea being that instead of a single `processElement` call having a number of ring ranges to query (but async), instead we'll have multiple calls to `processElement` each of which is a sync call. 

I agree that is probably most similar to the current behavior and I'm fine making that change.  Happy to hear other thoughts or suggestions.

aromanenko-dev commented:
> The idea being that instead of a single processElement call having a number of ring ranges to query (but async), instead we'll have multiple calls to processElement each of which is a sync call.

I agree with this change and it looks for me as a more `Beam`-ish way to do such things.

vmarquez commented:
@aromanenko-dev I've pushed the requested changes.  I also didn't rebase as I thought it'd be easier to review that way, I'm not sure what you prefer.

https://github.com/apache/beam/pull/10546/commits/e72e4d12094aa9fa8d80294b4f9d1617d50a13b3

Let me know and I can squash if you want.

aromanenko-dev commented:
@vmarquez Thanks! Sounds good - could you squash the commits with a proper commit message, including Jira ID, and rebase it against current `main` branch?

vmarquez commented:
@aromanenko-dev I squashed then rebased on master, added the JIRA to the commit msg.  Let me know how this looks an anything else I can do.

aromanenko-dev commented:
Run Java PreCommit

aromanenko-dev commented:
@vmarquez Please, take a look on `Java PreCommit` [job](https://ci-beam.apache.org/job/beam_PreCommit_Java_Phrase/4017/console) - it fails. `ElasticsearchIOTest.testSplit` fails are caused by flaky tests, so it's not caused by this PR. 

However, there are 2 other failing tasks that we need to fix before merging:
```
15:06:24 1: Task failed with an exception.
15:06:24 -----------
15:06:24 * What went wrong:
15:06:24 Execution failed for task ':sdks:java:io:cassandra:analyzeClassesDependencies'.
15:06:24 > Dependency analysis found issues.
15:06:24   usedUndeclaredArtifacts: 
15:06:24    - com.google.guava:guava:25.1-jre@jar

15:06:24 2: Task failed with an exception.
15:06:24 -----------
15:06:24 * What went wrong:
15:06:24 Execution failed for task ':sdks:java:io:cassandra:checkstyleMain'.
15:06:24 > Checkstyle rule violations were found. See the report at: file:///home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Phrase/src/sdks/java/io/cassandra/build/reports/checkstyle/main.html
15:06:24   Checkstyle files with violations: 1
15:06:24   Checkstyle violations by severity: [error:1]
```

aromanenko-dev commented:
Run Java PreCommit

aromanenko-dev reviewed (approved):
@vmarquez Thanks for working on this and finishing this PR! 
LGTM

vmarquez commented:
@aromanenko-dev thanks so much for your help, I hope I can continue to contribute!
